### PR TITLE
CI: End-to-end instance launch

### PR DESCRIPTION
**Pull request by iliana**
As demoed Friday, here's a test that launches a Debian instance in lab CI and successfully SSHes into it. There's hopefully enough infrastructure here to enable others to write lab tests as well; some things could definitely be moved into `mod helpers` but I'd like to avoid trying to predict what needs to be moved into helper functions.

Please also review https://github.com/oxidecomputer/sodiumoxide/commit/c2546d157fe62f762300b129a2adcfdfb3bb0503, which are changes to libsodium-sys's build.rs to support illumos. Once this lands I will ~~try to send that upstream so we can stop maintaining a fork~~ (welp, they archived the upstream repo). (libsodium-sys is a required component of thrussh, the library we are using as an SSH client in these tests so that we don't have to instrument the OpenSSH client for tests.)

### CODE REVIEW

File: end-to-end-tests/src/helpers/mod.rs
Comment (davepacheco) [inline end-to-end-tests/src/helpers/mod.rs:15]: You might find `test-utils::dev::poll::wait_for_condition` useful here?

**Review by davepacheco — COMMENTED** [review]
Cool!

I haven't taken a close look at the test but I like the approach of using a non-default workspace crate for running these tests.  It's been in the back of my mind how we'd run more complicated integration tests like this in a way that didn't trip up a standard `cargo test`, but were still well integrated into the workflows and tooling that people expect (i.e., still runnable by `cargo test`).  This looks like a good way to do that.

File: end-to-end-tests/src/helpers/mod.rs
Comment (iliana) [reply end-to-end-tests/src/helpers/mod.rs:15]: Oh, yep, this is basically what I wrote. I'll defer to the existing one :)

File: end-to-end-tests/src/bin/bootstrap.rs
Comment (luqmana) [inline end-to-end-tests/src/bin/bootstrap.rs:40]: Carving some space out of the same subnet as Nexus seems alright for now. Besides Nexus (and the global zone) was there anything else we'd conflict with?

File: sled-agent/src/bootstrap/params.rs
Comment (luqmana) [inline sled-agent/src/bootstrap/params.rs:34]: 😞

**Review by luqmana — APPROVED** [review]
Sweet this is awesome! It'll be great to have actual end-to-end tests now in addition to the sled-agent-sim ones! Left a couple small comments but otherwise looks good to me

File: end-to-end-tests/src/bin/bootstrap.rs
Comment (iliana) [reply end-to-end-tests/src/bin/bootstrap.rs:40]: I'm mostly worried about home network setups for people running these tests locally. I don't think there's a whole lot I can do there though.

(I've idly been wondering if we should have an out-of-repo config we look for describing a user's network setup.)

**Comment by iliana** [comment]
Before I merge, I'm looking into why, after a merge, the tests are now failing locally (bootstrap is timing out) and [in CI](https://github.com/oxidecomputer/omicron/runs/8097880939).

File: end-to-end-tests/src/bin/bootstrap.rs
Comment (luqmana) [reply end-to-end-tests/src/bin/bootstrap.rs:40]: > (I've idly been wondering if we should have an out-of-repo config we look for describing a user's network setup.)

Yes! I think that or even just something we stick in the .gitignore. Since the configs are part the repo today any modifications you make for your local testing (e.g. mac gateway) pollute git status.

**Comment by zephraph** [comment]
I've been working on a series of end-to-end tests which can drive the console against a live instance of Nexus. Once this is in, I'd like to setup a GitHub action to drive those tests against these instances. 

It's a bit of a long-tail in terms of CI and not something you'd want to wait on for development. I think it'd be helpful to run it against main at least.

**Comment by iliana** [comment]
That could be interesting — ultimately the long tail in this current test is the amount of time it takes to [build and package the control plane (22 min in this run)](https://buildomat.eng.oxide.computer/wg/0/details/01GFWB3D4Q37AH92QAJ8M9CCH5/9sPocl6Xr1yHHUx1B4X0phxmhktvr9XqsjjT11PtEvngI6CR/01GFWB3PE82GHNW26VW9NQBSWG), and then [about 5 minutes to deploy the control plane on a lab machine](https://buildomat.eng.oxide.computer/wg/0/details/01GFWB3D4Q37AH92QAJ8M9CCH5/9sPocl6Xr1yHHUx1B4X0phxmhktvr9XqsjjT11PtEvngI6CR/01GFWB3PDF39QQ2GKKQXSBGXX7). We do _have to_ stand up the control plane on a lab machine, of which we only have two currently, so there's not a lot of concurrency there; but if we do want to run a separate lab machine that is an optional CI step to run the console end-to-end tests, we certainly could in theory.

File: Cargo.toml
Comment (smklein) [inline Cargo.toml:121]: We chatted about this off github, but for posterity's sake:

- We wanted libsodium because it's a dependency of thrussh
- We use thrussh in the end-to-end test
- Thrussh also looks deprecated

We should at least file an issue to follow-up here, and minimize our use of unsupported deps.

File: end-to-end-tests/src/bin/bootstrap.rs
Comment (smklein) [inline end-to-end-tests/src/bin/bootstrap.rs:42]: Nit, can we at least move this "IP range fro the pool" construction to the "helpers"?

This seems like it'll definitely need to be changed as we continue updating how nexus gets launched, but it should probably be coupled tightly with "what network setup we used for nexus.

File: end-to-end-tests/src/instance_launch.rs
Comment (smklein) [inline end-to-end-tests/src/instance_launch.rs:224]: We *should* be deleting the instance before the project / org.

We aren't checking for this yet, but we will be in the future

**Review by smklein — APPROVED** [review]

File: end-to-end-tests/src/bin/bootstrap.rs
Comment (iliana) [reply end-to-end-tests/src/bin/bootstrap.rs:42]: Move it to the end-to-end helpers so that other test code could potentially get the first/last address in the pool that was set up? Or another use case (or somewhere else entirely)?

I can do this but I'm curious about why it would be necessary -- a test could also list the system IP pools if needed, which would probably be more reliable than parsing a config file that can potentially change / go away.

Comment (smklein) [reply end-to-end-tests/src/bin/bootstrap.rs:42]: My recommendation here was really small - basically just to move this to a function called `get_system_ip_pool` or something, so that the hack is isolated and can easily be replaced. I had no opinion on re-using this function within test bodies themselves -- that didn't seem necessary yet.

File: .github/buildomat/jobs/deploy.sh
Comment (iliana) [reply .github/buildomat/jobs/deploy.sh:131]: For the moment I'm avoiding this because I'd rather have the tcpproxy binary and its service manifest in sync in the same place. (But I'm making a note to eventually build tcpproxy in CI perhaps?)

**Pull request merged**
